indexing-service/src/main/java/io/druid/indexing/common/IndexTaskClient.java

```diff
@@ -0,0 +1,381 @@
+  {
+    new Socket(host, port).close();
+  }
+
+  protected FullResponseHolder submitRequestWithEmptyContent(
+      String taskId,
+      HttpMethod method,
+      String pathSuffix,
+      @Nullable String query,
+      boolean retry
+  ) throws IOException, ChannelException, NoTaskLocationException
+  {
+    return submitRequest(taskId, null, method, pathSuffix, query, new byte[0], retry);
+  }
+
+  /**
+   * To use this method, {@link #objectMapper} should be a jsonMapper.
+   */
+  protected FullResponseHolder submitJsonRequest(
+      String taskId,
+      HttpMethod method,
+      String pathSuffix,
+      @Nullable String query,
+      byte[] content,
+      boolean retry
+  ) throws IOException, ChannelException, NoTaskLocationException
+  {
+    return submitRequest(taskId, MediaType.APPLICATION_JSON, method, pathSuffix, query, content, retry);
+  }
+
+  /**
+   * To use this method, {@link #objectMapper} should be a smileMapper.
+   */
+  protected FullResponseHolder submitSmileRequest(
+      String taskId,
+      HttpMethod method,
+      String pathSuffix,
+      @Nullable String query,
+      byte[] content,
+      boolean retry
+  ) throws IOException, ChannelException, NoTaskLocationException
+  {
+    return submitRequest(taskId, SmileMediaTypes.APPLICATION_JACKSON_SMILE, method, pathSuffix, query, content, retry);
+  }
+
+  /**
+   * Sends an HTTP request to the task of the specified {@code taskId} and returns a response if it succeeded.
+   */
+  private FullResponseHolder submitRequest(
+      String taskId,
+      @Nullable String mediaType, // nullable if content is empty
+      HttpMethod method,
+      String pathSuffix,
+      @Nullable String query,
+      byte[] content,
+      boolean retry
+  ) throws IOException, ChannelException, NoTaskLocationException
+  {
+    final RetryPolicy retryPolicy = retryPolicyFactory.makeRetryPolicy();
+
+    while (true) {
+      FullResponseHolder response = null;
+      Request request = null;
+      TaskLocation location = TaskLocation.unknown();
+      String path = StringUtils.format("%s/%s/%s", BASE_PATH, taskId, pathSuffix);
+
+      Optional<TaskStatus> status = taskInfoProvider.getTaskStatus(taskId);
+      if (!status.isPresent() || !status.get().isRunnable()) {
+        throw new TaskNotRunnableException(StringUtils.format(
+            "Aborting request because task [%s] is not runnable",
+            taskId
+        ));
+      }
+
+      String host = location.getHost();
+      String scheme = "";
+      int port = -1;
+
+      try {
+        location = taskInfoProvider.getTaskLocation(taskId);
+        if (location.equals(TaskLocation.unknown())) {
+          throw new NoTaskLocationException(StringUtils.format("No TaskLocation available for task [%s]", taskId));
+        }
+
+        host = location.getHost();
+        scheme = location.getTlsPort() >= 0 ? "https" : "http";
+        port = location.getTlsPort() >= 0 ? location.getTlsPort() : location.getPort();
```

Comment:
Hmm, I really didn't write this class. This line was added in https://github.com/apache/incubator-druid/pull/4270 and was originally in `KafkaIndexingTaskClient`. I extracted some common parts from `KafkaIndexTaskClient` to `IndexTaskClient`, so that `KafkaIndexTaskClient` and `ParallelIndexTaskClient`. Please check https://github.com/apache/incubator-druid/pull/5492#discussion_r198956260 as well.

I think it's better to fix this class in a separate PR if needed because this PR is already big and I don't want to add unrelated changes.

Reply:
sounds good